### PR TITLE
improved tags loading performance, changed string hash algorithm to djb2

### PR DESCRIPTION
**Pull request by mrbald**

### CODE REVIEW

File: util/refString.c [hunk 1]
```diff
@@ -123,29 +111,14 @@ const char *RefStringDup(const char *str)
...
-            exit(EXIT_FAILURE);
-        }
-        strcpy(newrcs->string, str);
+        struct rcs* newrcs = (struct rcs*) NEditMalloc(sizeof(struct rcs));
         newrcs->usage = 1;
-        newrcs->next = NULL;
+        newrcs->next = Rcs[bucket];
+        Rcs[bucket] = newrcs;
+
+        newrcs->string = (char*) NEditMalloc(len + 1);
```
Comment (ajbj) [inline util/refString.c]: Ideally a `NEditStrndup(const char *, size_t)` would be better here. (`NEditStrdup` would ideally use `NEditStrndup` too.)

File: util/refString.c [hunk 2]
```diff
@@ -123,29 +111,14 @@ const char *RefStringDup(const char *str)
...
-        }
-
-        if (NULL == (newrcs->string = (char*) malloc(len + 1))) {
-            /*  Not much to fall back to here.  */
-            fprintf(stderr, "nedit: rcs_strdup(): out of heap space!\n");
-/*            XBell(TheDisplay, 0); */
-            exit(EXIT_FAILURE);
-        }
-        strcpy(newrcs->string, str);
+        struct rcs* newrcs = (struct rcs*) NEditMalloc(sizeof(struct rcs));
```
Comment (ajbj) [inline util/refString.c]: `struct rcs *newrcs = NEditNew(struct rcs)`?

File: util/refString.c [hunk 3]
```diff
@@ -50,21 +51,31 @@ struct rcs
 static struct rcs       *Rcs[RCS_SIZE];
 static struct rcs_stats  RcsStats;
 
-/*      Compute hash address from a string key */
+static unsigned const DJB2_SEED = 5381;
```
Comment (ajbj) [inline util/refString.c]: `5381u`?

File: util/refString.c [hunk 4]
```diff
@@ -50,21 +51,31 @@ struct rcs
...
-        x += a;
-    }
-    
-    for (a=1; i<(s+1); i++, a *= 256)
-        x += key[i] * a;
-        
-    return x;
+    unsigned hash = DJB2_SEED;
+    hash = djb2s_(key, hash);
+    return hash;
```
Comment (ajbj) [inline util/refString.c]: One liner: `return djb2s_(key, DJB2_SEED);`?

File: util/refString.c [hunk 5]
```diff
@@ -50,21 +51,31 @@ struct rcs
 static struct rcs       *Rcs[RCS_SIZE];
 static struct rcs_stats  RcsStats;
 
-/*      Compute hash address from a string key */
+static unsigned const DJB2_SEED = 5381;
+/* djb2s hash for null-terminated string, seeded version */
+unsigned djb2s_(char const* key, unsigned hash)
```
Comment (ajbj) [inline util/refString.c]: static function perhaps?

File: util/refString.c (hunk 3 above)
Comment (ajbj) [inline util/refString.c]: `5381u`?

File: source/tags.c [hunk 6]
```diff
@@ -206,6 +205,76 @@ static tag *getTag(const char *name, int search_type)
...
+
+    return NULL;
+}
+
+/*
+** Pre-loads tags into a temporary hash table with composite key made of
+** `name`, `file`, and `search` arguments.
+** The `rehashPreloadedTags()`, invoked after all tags are preloaded,
+** is rearranging tags into `Tags` hash table using `name` as the key.
+*/
```
Comment (ajbj) [inline source/tags.c]: Does this preload (presumably there to avoid collisions) make much of a difference in load times?

Comment (mrbald) [inline source/tags.c]: Thanks for asking, it's actually the pre-interning that makes a huge difference. I first added the collition resolution code and did not benchmark which of two is the major contributor. I'll back off the two-pass hashing.

File: util/refString.c (hunk 1 above)
Comment (mrbald) [inline util/refString.c]: Is there an `NEditStrndup`?

Comment (ajbj) [inline util/refString.c]: not yet!

**Pull request merged**